rfd/0133-connect-my-computer.md

```diff
@@ -0,0 +1,513 @@
+and to open the status tab. The status tab displays errors related to starting or stopping the
+agent. It also lets the user remove the agent completely.
+
+### Initial setup
+
+Let’s assume that we have a Team plan cluster at acme.teleport.sh with a user robert@acme.com.
+robert@acme.com uses bob as the system username on their computer.
+
+Once the user begins the setup, Connect performs all steps without further user interaction. Each
+step can independently fail with an error message explaining what happened. The setup process can be
+retried on error.
+
+#### Setting up roles
+
+
+The setup starts with the creation of the relevant role. If the user does not have permissions to
+create and set roles, this step fails with an error message explaining which permissions are needed.
+
+The role provides the necessary system login and access privileges. This way we have a single role
+providing both rather than a role which provides just the login and then CMC having to depend on
+another role, such as the built-in access, to provide access privileges.
+
+The role is scoped per cluster user. Its name follows the format of
+`connect-my-computer-<cluster-username>`. This is to accommodate multiple cluster users with the
+same system username wanting to use CMC – their CMC roles will not conflict with each other.
+
+For our example user, the YAML representation of the role looks like this:
+
+```yaml
+kind: role
+version: v6
+metadata:
+  name: connect-my-computer-robert@acme.com
+spec:
+  allow:
+  node_labels:
+    teleport.dev/connect-my-computer/owner: robert@acme.com
+  logins: [bob]
+```
+
+Connect creates this role and adds it to the current user. Since the user cert in Connect at the
+time does not contain the new role, Connect has to refresh the roles to enable the user to connect
+to the new node later on. This is done by calling `TeleportClient.ReissueUserCerts` with
+`client.CertCacheDrop` as the second argument since [we expect the roles to change during the reissue](https://github.com/gravitational/teleport/blob/60830b0d8a7d5e253bd00b0967929ddb913b9b4a/lib/client/client.go#L290-L293).
+
+#### Downloading the agent
+
+
+To download the teleport binary, Connect needs to construct a URL of the tar.gz file with binaries.
+The URL contains the OS, the architecture and the agent version. The OS and the architecture are
+already kept in `RuntimeSettings`.
+
+For Linux, the downloaded file is one of:
+* `teleport-v<version>-linux-arm64-bin.tar.gz`
+   * We don’t have Connect arm64 releases, but it will be useful for testing on VMs.
+* `teleport-v<version>-linux-amd64-bin.tar.gz`
+
+For macOS:
+* `teleport-v<version>-darwin-arm64-bin.tar.gz`
+* `teleport-v<version>-darwin-amd64-bin.tar.gz`
+
+The MVP will not verify the checksum of the downloaded archive.
+
+##### Agent versioning
+
+The version of the agent that Connect uses always matches the version of Connect. If Connect detects
```

Comment:
Again, to keep things simple I think we should just always go with the matching version. Adding a check for the latest compatible patch would require implementing the logic for this check (I suppose we could use `https://rlz.teleport.sh/teleport?majorVersion=12` for that) and then deciding when to perform it and offering a way to run it manually.

By tying the agent version to the Connect version, it's super easy to explain the update process of the agent. The user loses the ability to have Connect version X.Y.Z and running an agent X.Y.Z+n but I don't see why someone would not want to also upgrade Connect in this situation.

---

This made me realize that in the RFD we didn't account for the fact that if Connect version X adds a node to cluster version X+1, the user will see a warning in the Web UI about outdated agents but Connect will be silent about it.

Since we're going to perform version checks anyway, maybe we should also nudge the user towards upgrading the agent (by updating Connect) in that situation.

Reply:
I expanded the "Agent versioning" section with notes on how the UI should handle version differences, including a situation where connecting an agent through Connect is going to show the "Some Agents within this Cluster are running an older version of Teleport" message in the Web UI.

Reply:
> it's super easy to explain the update process of the agent

What exactly is the upgrade process of the agent? Upgrade Teleport Connect and reactivate the CMC flow in some way?

Reply:
cc @gzdunek @roraback on how to approach the upgrade process of an agent.

I should probably explain the upgrade process in a more explicit way in the RFD [than it is right now](https://github.com/gravitational/teleport/blob/rfd/0133-connect-my-computer/rfd/0133-connect-my-computer.md#keeping-the-agent-up-to-date).

We planned for the upgrade process to look like this:

1) The user is logged in to a cluster and closes Connect 13.0.0 without manually stopping Connect My Computer.
2) The user upgrades Connect to 13.0.1 and launches it again.
3) Since Connect My Computer was not manually stopped and the user has a valid cert, Connect attempts to start the agent soon after the app starts.
4) Before actually starting the agent, Connect sees that the agent version 13.0.1 is missing from the cache.
5) Connect downloads the agent 13.0.1.
6) The new agent is launched as usual.

Back when Grzegorz and I were discussing this, we didn't account for the macOS system dialog that is shown before launching a downloaded binary for the first time.

If we leave it as it is, after upgrading Connect the user will see the dialog:

> “teleport” is an app downloaded from the Internet. Are you sure you want to open it?

They'll likely have no idea what is going on as teleport was launched without their interaction. Downloading the tar.gz with binaries might also take some time, so worst case scenario they just see that dialog out of nowhere after two or three minutes. I'd be scared in that situation, especially since Teleport is supposed to provide secure access to the infrastructure and now I see a system dialog about teleport that I have no idea where it came from. I'd assume I might be getting attacked by someone. 🫨

---

We could salvage this by showing some in-app notifications before & after download, but given that there might be a significant delay in time before the download is started and the dialog is shown, I think it might be better to not download an agent automatically if it's missing from the cache.

Instead, we could add a special state for the CMC, something like `agent-missing`. The UI could then show some sort of a warning icon next to the CMC UI, wherever the UI ends up being placed, and the status tab could explain that the agent was not started automatically because the agent binary is missing, most likely due to Teleport Connect upgrade or being removed from the cache.

The user would need to start the agent manually, which would then actually download the new agent.

The `agent-missing` state would be set only during autostart of the agent.

Reply:
>“teleport” is an app downloaded from the Internet. Are you sure you want to open it?

Does this happen when a binary is spawned using Node?

I just did a test, I download teleport and did not touch it, then added a code that spawns a new process and packaged the app. The only dialog I saw was about access to Download folder (because `teleport` was ran from there).
Did you also test this in a similar way and have different results?

Reply:
I just downloaded Transmission and it did ask me before running the app for the first time. 🤔

Reply:
Isn't it related to `.app` files vs regular binaries?
It seems to me that I have never seen the dialog when running CLI binary.

But I don't know how to explain the issue you had three days ago. You didn't spawn `tsh.app`, right?

Reply:
I talked briefly with Grzegorz about this, to sum this up:

* We don't know why the behavior is different now vs three days ago.
* My tests from three days ago were insufficient since I didn't account for [the quarantine flag](https://eclecticlight.co/2020/10/29/quarantine-and-the-quarantine-flag/). The browsers, I assume, always set the quarantine flag for all downloaded files, whereas that doesn't have to be the case for our app (see the article).
* I also found [Enabling User-Selected File Access](https://developer.apple.com/library/archive/documentation/Miscellaneous/Reference/EntitlementKeyReference/Chapters/EnablingAppSandbox.html#//apple_ref/doc/uid/TP40011195-CH4-SW6) in [a Developer Forums thread](https://developer.apple.com/forums/thread/683551).
   * Keep in mind this is about sandboxed App Store apps whereas our app is not sandboxed AFAIR, so that entitlement might not apply to us.

@gzdunek Do you have the time to write the prototype you mentioned? Something that would download an archive binary from a hardcoded URL, run `teleport version` and then output it somewhere in the app? As you suggested earlier in DMs, we could make a tag build out of that and get some actual answers.

Reply:
>Keep in mind this is about sandboxed App Store apps whereas our app is not sandboxed AFAIR, so that entitlement might not apply to us.

Correct, we are not sandboxed.

Yes, I can work on the prototype, I will share results tomorrow.

Reply:
I finished the prototype and tested it.
Good news:
- I ran the app on a fresh macOS VM - the only dialog I saw was about running Connect, `teleport version` started with no issues.

Other observations:
- The binary has a `com.apple.provenance` attribute. I think we have not seen it before. It is added when quarantine is cleared, starting from Ventura. Someone reported that:
>I installed iTerm2 and noticed that provenance being added to every directory I create from it and not being able to remove even after I gave iTerm2 FDA. But if I launch Terminal (which I had given FDA), I can delete the attribute.
So it seem iTerm2 has been removed from quarantine but Apple wants to track everything that gets created using it. That is going to be a lot of noise in my filesystem.

I was afraid it may cause some problems related to Gatekeeper in the future, but I see that the attribute is added to every file that app creates (like `app_config.json`), so there is nothing we can do about it anyway.

Reply:
Thanks for testing this out.

I noticed that the .tar.gz also has `com.apple.macl` set.

* https://eclecticlight.co/2020/01/30/quarantine-sip-and-macl-macos-per-file-security-controls/
* https://eclecticlight.co/2023/05/10/how-macos-now-tracks-the-provenance-of-apps/
* https://eclecticlight.co/2023/05/18/inspect-and-control-extended-attributes-with-xattred-1-4/
* https://blog.xpnsec.com/we-need-to-talk-about-macl/

The tl;dr is that it most probably signifies that Connect downloaded the file and allows it to edit it.

I'm glad to see that it's working and, at least as it appears for now, we don't have to worry about the Gatekeeper prompt. This means that the setup will remain interaction-free.

I'm still not sure about the UX of the update process. Despite not having to worry about the prompt, I don't know if we should download the new agent automatically or direct the user towards the status page [as I originally described](https://github.com/gravitational/teleport/pull/27815#discussion_r1238386004). That's something we should talk about with Kenny.

---

@gzdunek Could you also timebox an hour or two to research the `com.apple.security.files.user-selected.executable` entitlement that [I mentioned](https://github.com/gravitational/teleport/pull/27815#discussion_r1238683347)? I'd start with the [Enabling User-Selected File Access](https://developer.apple.com/library/archive/documentation/Miscellaneous/Reference/EntitlementKeyReference/Chapters/EnablingAppSandbox.html#//apple_ref/doc/uid/TP40011195-CH4-SW6) section I linked to in my comment. I'd like us to be sure that:

* We follow best practices set out by Apple.
* If the entitlement is needed only for sandboxed apps or if developer-signed apps also need to account for it.

It might be the case that Electron/electron-builder asks for that entitlement OOTB.